### ISSUE TITLE
javax.net.ssl.SSLException: SSLEngine is closing/closed status

### ISSUE THREAD

**Issue by psantacl** [http-kit/http-kit#324]
On java 7 OpenJDK we have enabled TLS1.2 as follows: 

```
(def ssl-engine
  (let [tm (reify javax.net.ssl.X509TrustManager
             (getAcceptedIssuers [this] (make-array X509Certificate 0))
             (checkClientTrusted [this chain auth-type])
             (checkServerTrusted [this chain auth-type]))
        client-context (SSLContext/getInstance "TLSv1.2")]
    (.init client-context nil
           (-> (make-array TrustManager 1)
               (doto (aset 0 tm)))
           nil)
    (.createSSLEngine client-context)))

```
the ssl-engine is then passed as part of the request options to the http-client: 
```
 @(http/request {:timeout    request-timeout
                           :keep-alive request-keep-alive
                           :headers    (base-headers)
                           :insecure?  false
                           **:sslengine ssl-engine**
                           :url "some-url"
                           :method :get })

```

After some random number of requests, the exception mentioned in the subject line is thrown and no further HTTP requests will succeed until the ssl-engine is recreated.   After doing some research, we uncovered the following fixed issue in a newer version of netty: https://github.com/netty/netty/commit/9efe48fc3a3204f8808cd1dc111ab6ee16aff6c8

Not sure if this relevant but it seems promising.

**Comment by ccann** [comment]
Is there any progress on this issue? A workaround perhaps?

**Comment by singen** [comment]
I'm also running into this exception in our production setup when experiencing over 10 requests per minute. We don't even have a custom SSLEngine, we just run the default context. The next request will usually succeed though, so the SSLEngine is (I guess) recreated.

**Comment by javahippie** [comment]
I would be happy to look into this issue. If somebody who runs into it could post a full stacktrace, or a small example application which reproduces the error, it would make things easier.

**Comment by ptaoussanis** [comment]
Thanks Tim, much appreciated! @singen, @ccann, @psantacl: any possibility someone could provide a full stacktrace and/or small reproducible example to help Tim with debugging?

**Comment by olessavluk** [comment]
I've received this error, but it happened after the first unsuccessfull request (which is maybe normal)
@javahippie I've creadet another issue with an example app, see https://github.com/http-kit/http-kit/issues/334#issuecomment-298238178

**Comment by singen** [comment]
Hi, here is my stack trace from when the error occurs. Unfortunately I do not have time to make a reproducible example, but afaik there is nothing odd about how we are using the http-kit client.
It's just a make-query call with https urls.

```
:cause SSLEngine is closing/closed
:via
[{:type javax.net.ssl.SSLException
:message SSLEngine is closing/closed
:at [sun.security.ssl.SSLEngineImpl kickstartHandshake SSLEngineImpl.java 713]}]
:trace
[[sun.security.ssl.SSLEngineImpl kickstartHandshake SSLEngineImpl.java 713]
[sun.security.ssl.SSLEngineImpl beginHandshake SSLEngineImpl.java 756]
[org.httpkit.client.HttpClient finishConnect HttpClient.java 347]
[org.httpkit.client.HttpClient run HttpClient.java 424]
[java.lang.Thread run Thread.java 745]]}
```

**Comment by supritha** [comment]
Had the same issue when trying to make http request to an external system
:cause "SSLEngine is closing/closed"
 :via
 [{:type javax.net.ssl.SSLException
   :message "SSLEngine is closing/closed"
   :at [sun.security.ssl.SSLEngineImpl kickstartHandshake "SSLEngineImpl.java" 713]}]
 :trace
 [[sun.security.ssl.SSLEngineImpl kickstartHandshake "SSLEngineImpl.java" 713]
  [sun.security.ssl.SSLEngineImpl beginHandshake "SSLEngineImpl.java" 756]
  [org.httpkit.client.HttpClient finishConnect "HttpClient.java" 310]
  [org.httpkit.client.HttpClient run "HttpClient.java" 375]
  [java.lang.Thread run "Thread.java" 748]]}})

Initially thought it must be a bug with the library but its not. My code was trying to fire request recursively to the external system crossing the threshold of tcp connection. After the threshold tcp connection is closed hence the error

**Comment by benbramley** [comment]
I ran into  this issue recently. In my case it is related to the idle connection timeout settings usually on a firewall/load balancer between the client and the web server but some web servers also have this setting as well. For example AWS ELB has a default idle timeout of 60 seconds which means if a connection has not had any data sent/received in 60 seconds the ELB terminates the connection. There is a period of time where the OS (in my case I have replicated this on AIX, Linux and OSX) thinks the connection is still working (until the keep alive probe runs). If you happen to try to send data within this period using the connection you will get the error:

```
:cause "Cannot kickstart, the connection is broken or closed"
 :via
 [{:type javax.net.ssl.SSLException
   :message "Cannot kickstart, the connection is broken or closed"
   :at [sun.security.ssl.TransportContext kickstart "TransportContext.java" 209]}]
 :trace
 [[sun.security.ssl.TransportContext kickstart "TransportContext.java" 209]
  [sun.security.ssl.SSLEngineImpl beginHandshake "SSLEngineImpl.java" 97]
  [org.httpkit.client.HttpClient finishConnect "HttpClient.java" 412]
  [org.httpkit.client.HttpClient run "HttpClient.java" 492]
  [java.lang.Thread run "Thread.java" 748]]}
```
The error/stack trace is slightly different as I believe there have been some changes in more recent patches of JDK 8.

I was able to workaround it temporarily by reducing the keepalive setting from the default of 120 seconds down to 5 seconds, or turning off persistent connections also works (with the obvious downside of being more expensive).

I have not had time to go through the code in detail yet but I believe the problem stems from recycling a connection when the SSLEngine is in a closed state. The connection gets recycled fine, but the old SSLEngine is also re-used which then leads to this error.

**Comment by ptaoussanis** [comment]
For anyone posting follow-ups in here, please make sure to mention what version of http-kit you're running.
@benbramley Could you please add this info?

Note that v2.16.0 includes fixes for some issues that could be related, e.g.: #401 #489 

Thanks!

**Comment by benbramley** [comment]
@ptaoussanis apologies I have updated my comment with the version info

**Comment by benbramley** [comment]
@ptaoussanis - I took a look at #489 and it does indeed look like it should fix it, I just re-tested 2.6.0 by clearing all dependencies etc. and this time it did not recreate, I think 2.6.0 has fixed it

**Comment by ptaoussanis** [comment]
@benbramley No problem, thanks for the added info 👍 

That's unfortunate, was hoping this'd have been addressed with #489.

Next step would be for you or someone else to try provide a reproducible example with the latest http-kit version (currently v2.6.0) to help further the investigation.

**Comment by ptaoussanis** [comment]
@benbramley Seems we both posted at the same time. Okay then, will mark this issue as closed 👍 

If anyone else does run into the problem again - please try to provide a reproducible example with the latest http-kit version.

Thanks!
